dataprochub/app.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Replaces default JupyterHub.app to handle redirects. """
+
+from jupyterhub.app import JupyterHub
+from jupyterhub.handlers.base import UserUrlHandler
+
+class DataprocHubUserUrlHandler(UserUrlHandler):
+  """ Extends UserUrlHandler to redirect user once spawn is done. """
+
+  async def _redirect_to_user_server(self, user, spawner):
+    self.statsd.incr('redirects.user_after_login')
+    redirect_url = user.spawners[spawner.name].component_gateway_url
+    self.log.info(f'# Redirecting to notebook at {redirect_url}.')
+    self.redirect(redirect_url)
+
+class DataprocHub(JupyterHub):
+  """ Extends JupyterHub mainly to handle redirect vs proxy.
+
+  The order of the handlers is importat so replaces dynamically when relevant.
+  """
+
+  def __init__(self, *args, **kwargs):
+    super().__init__(*args, **kwargs)
+
+    self.new_user_handler = (
+      r'/user/(?P<user_name>[^/]+)(?P<user_path>/.*)?',
+      DataprocHubUserUrlHandler,
+    )
+
+    self.new_user_handler_prefixed = self.add_url_prefix(
+        self.hub_prefix, [self.new_user_handler])[0]
+
+    self.regex_to_change = {
+      r'%s(user|services)/([^/]+)' % self.base_url: r'%s(services)/([^/]+)' % self.base_url
+    }
+
+    self.handler_to_change = {
+      self.new_user_handler_prefixed[0]: self.new_user_handler_prefixed[1]
+    }
+
+  def init_handlers(self):
+    """ Modifies the default app handlers.
+
+    Modifies the default behavior of JupyterHub user URL handler to use a custom
+    handler and allow Dataproc Hub to do redirect instead of proxying.
+
+    Handlers are tuples of the form (regex_path, handler_class).
+
+    Order of the element in the handlers list matters!!!
+    """
+    super().init_handlers()
+
+    for idx, h in enumerate(self.handlers):
+      h_regex = h[0]
+      h_class = h[1]
+
+      if h_regex in self.regex_to_change:
```

Comment:
This makes me a little uncomfortable: this seems like it will be very brittle if they change future versions of JupyterHub.

I would prefer we don't have to mess with the handlers that are constructed by the base class at all: would it work to just put our handler first in the list, since they're always matched in order?

Alternatively, could we match by handler class instead of the path regex?

Reply:
Will check if inserting ours without deleting others work. 
If not, the class sounds good, I wasn't sure which of the class vs regex was the less likely to change in JupyterHub code

Reply:
Fixed in 394df2a35affb70481e80647c0e7e6e928938f1d